test/api/test_plan_serialization_bwc.cpp

```diff
@@ -0,0 +1,71 @@
+#include "catch.hpp"
+#include "duckdb/common/field_writer.hpp"
+#include "duckdb/common/serializer/buffered_file_reader.hpp"
+#include "duckdb/common/serializer/buffered_file_writer.hpp"
+#include "duckdb/planner/planner.hpp"
+#include "duckdb/parser/parser.hpp"
+#include "test_helpers.hpp"
+#include "tpch-extension.hpp"
+
+using namespace duckdb;
+using namespace std;
+
+string get_file_name(FileSystem &fs) {
+	auto file_name = string("test_plan_serialization_bwc.cpp");
+	auto path = string(__FILE__);
+	return path.replace(path.rfind(file_name), file_name.length(), "serialized_tpch_queries.binary");
+}
+
+TEST_CASE("Generate serialized TPCH file", "[.]") {
+	DuckDB db;
+	Connection con(db);
+	con.Query("CALL dbgen(sf=0.01)");
+	BufferedFileWriter serializer(db.GetFileSystem(), get_file_name(db.GetFileSystem()));
+	serializer.SetVersion(PLAN_SERIALIZATION_VERSION);
+	serializer.Write(serializer.GetVersion());
+	for (int q = 1; q < 22; q++) {
+		auto sql = TPCHExtension::GetQuery(q);
+
+		con.BeginTransaction();
+		Parser p;
+		p.ParseQuery(sql);
+
+		Planner planner(*con.context);
+
+		planner.CreatePlan(move(p.statements[0]));
+		auto plan = move(planner.plan);
+		plan->Serialize(serializer);
+
+		con.Rollback();
+	}
+	serializer.Sync();
+}
+
+TEST_CASE("Test deserialize TPCH from file", "[serialization]") {
+	DuckDB db;
+	Connection con(db);
+	con.Query("CALL dbgen(sf=0.01)");
+	BufferedFileReader deserializer(db.GetFileSystem(), get_file_name(db.GetFileSystem()).c_str());
+	deserializer.SetVersion(deserializer.Read<uint64_t>());
+
+	for (int q = 1; q < 22; q++) {
+		auto sql = TPCHExtension::GetQuery(q);
+
+		con.BeginTransaction();
+		Parser p;
+		p.ParseQuery(sql);
+
+		Planner planner(*con.context);
+
+		planner.CreatePlan(move(p.statements[0]));
+		auto plan = move(planner.plan);
+		PlanDeserializationState state(*con.context);
+		auto read_plan = LogicalOperator::Deserialize(deserializer, state);
+
+		//		printf("[%d] Original plan:\n%s\n", q, plan->ToString().c_str());
+		//		printf("[%d] New plan:\n%s\n", q, read_plan->ToString().c_str());
+
+		REQUIRE(read_plan->ToString() == plan->ToString());
```

Comment:
Perhaps running the plans and verifying that the correct result is returned is a more robust verification method?

Reply:
so that's a good a question that I was doubting about. Ideally you'd like some method to compare the query internals. Having the same results on a specific database doesn't guarantee the query is the same (naive example: a change in a LIMIT won't show on a database with less rows). Not sure how much that holds for the TPCH database.  I'm happy to follow what ever you think is best but that was my reasoning. Shall I add result based testing? Shall keep the ToString as pre-flight? Are we just good as is?

Reply:
That is true, and perhaps the `ToString` is good to keep in addition to running the query, but I would still run the query and verify that it produces the same result. There are many things that can go wrong with (de)serialization that won't show up in `ToString` as the `ToString` only prints a subset of the internals of each of the operators/expressions.

Reply:
clear. will do. Just to be clear, do you prefer keeping the to string implementation?

Reply:
Yes, having both is fine.